kernel/waker_generic/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+//! A standalone, generic implementation of a Waker + Blocker pair -- a basic wait event.
+//!
+//! By standalone, we mean that this crate does not depend on tasking infrastructure
+//! (e.g., the `task` crate) directly, meaning it is usable in the `task` crate.
+//! This works by accepts generic actions (closures) for waking the [Waker](core::task::Waker)
+//! and blocking the [Blocker].
+
+#![no_std]
+
+extern crate alloc;
+
+use alloc::sync::Arc;
+// @tsoutman: as I said before, I think we definitely do need a preemption-safe mutex here.
+//            It's not technically required, but it'd prevent task switch thrashing
+//            between the task that is being blocked and the task that is waking the waker.
```

Comment:
We can benchmark this, but I don't think we should use preemption-safe mutexes when they aren't strictly required. Disabling the APIC takes time and also prevents tasks with a higher priority from running.

Disabling preemption is also useless in the likely scenario that the waker task is running on another core.

Reply:
Fair enough. The one concern i had is demonstrated by this sequence below
```rust
let mut woken = self.woken.lock();
if *woken {
    *woken = false;
    return;
} else {
    let r = block_action(); // blocks the current thread
    // <-- preemptive task switch occurs here, before `woken` lock can be dropped
    drop(woken);
    drop(r);
}
```
if a preemptive task switch occurs where indicated by my comment, the lock on `woken` will not be dropped. 
Say the "waker" task then runs and attempts to wake the associated waker, it will not be able to obtain the lock on its inner `woken` field and will just have to spin endlessly until the "blocker" task gets to run again and release its lock.

Disabling preemption here may indeed be too expensive, but it would at least prevent that scenario.

Perhaps there's another better solution -- just dropping the lock on `woken` right before invoking the `block_action()`. I think that wouldn't cause potential interleaving issues with the waker, so long as we keep the `wake()` function as is... but I haven't thought deeply about it. Let me know what you think.

Reply:
> I think that wouldn't cause potential interleaving issues with the waker, so long as we keep the wake() function as is... but I haven't thought deeply about it.
I think it would in the following order:
1. Blocker drops the lock and is preempted
2. Waker wakes the blocker (unblocking an already unblocked task)
3. Blocker is rescheduled and blocks itself, but won't ever be unblocked.

hm, yea. then we probably do need to disable preemption for correctness.

Reply:
so, to be clear, the problematic interleaving would only occur if we did drop the lock on `woken` _before_ invoking `block_action()`. It wouldn't occur in the current code.

Reply:
One alternative idea is to add a new function that cheaply & quickly disables preemption function (without messing with interrupt chip timers). So it'd be appropriate for this scenario and all other scenarios where we want to atomically block a task and then do something else without the scheduler interfering.

Reply:
Yea I was thinking about that as well. I'm not sure how long disabling the APIC takes, but I think it would be faster to soft disable preemption (continue ticks) if it's just for one scheduler invocation.

As a side note, I think Theseus preempts much more often than other operating systems so that's something we could adjust.

Reply:
So after more thought, we can't reorder the blocking operation and lock release, even if we use preemptive-safe locks. That's fine.

> Yea I was thinking about that as well. I'm not sure how long disabling the APIC takes, but I think it would be faster to soft disable preemption (continue ticks) if it's just for one scheduler invocation.

I don't think it's expensive, but it does require acquiring yet another lock. I can experiment with that disabling preemption during the blocking procedure in a future PR, since it doesn't affect correctness, only potential performance. 

> As a side note, I think Theseus preempts much more often than other operating systems so that's something we could adjust.

Do we? I think our simple preemptive scheduling timer interrupt occurs every 6ms or so, which I thought was on the lower side of things -- most commercial OSes vary their timeslice periods from 0.5 to 8ms, but in any case I'm happy to change it if we determine it'd be valuable.